runtime/vm/classsupport.c

```diff
@@ -1067,8 +1160,14 @@ loadNonArrayClass(J9VMThread* vmThread, J9Module *j9module, U_8* className, UDAT
 				/* check again if somebody else already loaded the class */
 				foundClass = hashClassTableAt(classLoader, className, classNameLength);
 				if (NULL != foundClass) {
-					omrthread_monitor_exit(vm->classTableMutex);
-					goto done;
+					if (!IS_RESTORE_RUN(vm)) {
+						omrthread_monitor_exit(vm->classTableMutex);
+						goto done;
+#if defined(J9VM_OPT_SNAPSHOTS)
+					} else if (!loadWarmClassFromSnapshot(vmThread, classLoader, foundClass)) {
+						foundClass = NULL;
+#endif /* defined(J9VM_OPT_SNAPSHOTS) */
+					}
```

Comment:
When `J9VM_OPT_SNAPSHOTS` is involved, I think uses of `IS_RESTORE_RUN()` should be guarded.